Add randomized allocation sampling

chrisnas opened the pull request:
This PR adds a new feature, randomized allocation sampling, that can be used for low overhead allocation sampling with good probabilistic error bounds. We expect APM tools and other scenarios that do memory profiling in development or production will be interested in this. See [docs/design/features/RandomizedAllocationSampling.md](https://github.com/dotnet/runtime/pull/100356/files#diff-f0760b2f7659c7433b5313f11d30bf17f0061252058af8511ad117bfee99ceb3) for more details about what this is and how it works.

This PR is a continuation from work originally in https://github.com/dotnet/runtime/pull/98167. @Maoni0 and @jkotas had some discussion at that point to agree on the overall development direction.

There are a variety of test results showing some performance comparisons and statistical sampling distributions in src/tests/tracing/eventpipe/randomizedallocationsampling/manual/testing_results/. These are available for reviewers to see but will be deleted prior to merging the PR.

Current status: 
- The CoreCLR support is mostly complete but still tracking down some CI failures now that the code is being built and tested in a wider variety of configs
- NativeAOT support is in progress: the new event is emitted (currently looking for a better randomizer + measure statistical distribution but more complicated since the name of the type is not known)

Note:
An option could be to merge the Core part now to secure the feature in .NET 9 and continue the NativeAOT that might not be finished for the cut. Note that the fix to emit AllocationTick in NativeAOT is now merged so it could help if the new randomized sampling is not available.

Guide to the changes:
- The majority of changed files in the product code are simple refactoring to accommodate the new ee_alloc_context struct. This struct wraps a gc_alloc_context and adds one additional field combined_limit. All of the fast allocation helpers that previously referred to alloc_limit now refer to combined_limit.
- A few more interesting changes are:
  - src/coreclr/vm/gcheaputilities.h - Adds the definition of ee_alloc_context which tracks the state of randomized sampling and generates new random numbers for the sampler.
  - src/coreclr/vm/gchelpers.cpp - Modifications of the Alloc() method check if sampling is enabled, check if a new allocation should be sampled, and fire the sampling event if so. There is also a separate change AllocateSzArray() that simplifies aligning double arrays and removes a potentially unnecessary min_obj padding allocation there.
  - src/coreclr/vm/gcenv.ee.cpp - Modifies GCToEEInterface::GcEnumAllocContexts() to ensure that the combined_limit field is kept up-to-date with potential alloc_context changes made by the GC.
  - src/coreclr/gc/gc.cpp - The GC_ALLOC_ALIGN8 is now always supported. Previously it was only supported when FEATURE_64BIT_ALIGNMENT was enabled but we wanted to use it more broadly.

noahfalk committed:
This is an incomplete implementation of a new randomized allocation sampling feature. This commit is

only to gather feedback on the direction and is not intended to be checked in.

The is a resumption of some of the original work that occured in https://github.com/dotnet/runtime/pull/85750
although this is a different implementation approach. The overall goal is to do lightweight allocation
sampling that can produce unbiased approximations of what was actually allocated. Our current ETW AllocationTick
sampling is periodic but not randomized. Extrapolating from the AllocationTick samples can lead to unbounded
estimation errors in theory and also substantial estimation error observed in practice.

This commit is primarily to get feedback on adding a new adjustable limit pointer on
the coreclr EE side of allocation contexts. This allows breaking out of the fast path allocation helpers at an
arbitrarily selected sampling point that need not align with the standard GC allocation
quantum for SOH. With sampling off the fast_helper_alloc_limit would always be
identical to alloc_limit, but when sampling is on we would occasionally have them
diverge.

The key changes are:
gcheaputilities.h - defines a new ee_alloc_context which augments the existing gc_alloc_context with the extra fast helper limit field
gcheaputilities.h and threads.h - replace the global and per-thread storage of gc_alloc_context with the expanded ee_alloc_context
  to store the new field.
jithelpers.cpp, jitinterfacex86.cpp, and amd64 asm stuff - refactors fast path alloc helpers to use the new limit field instead
gchelpers.cpp - Updated Alloc() function recognizes when fast helper limit was exceeded to do some kind of sampling callback
  and update the fast helper limit when needed.

This commit doesn't contain any logic that actually turns the sampling on, determines the limit values needed for sampling,
or issues callbacks for sampled objects. That would come later if folks think this is a reasonable approach to intercept
allocations.

chrisnas committed:
Initial update

chrisnas committed:
Update based on feedback

chrisnas committed:
Rename fast_alloc_helper_limit_ptr to alloc_sampling

chrisnas committed:
Remove trailing whitespace in markdown

chrisnas committed:
Add sampling threshold computation

chrisnas committed:
Take feedback into account and start to implement AllocationSampled event

chrisnas committed:
Fix typo

chrisnas committed:
Take review into account

chrisnas committed:
Emit the AllocationSampled event

chrisnas committed:
Fix threshold computation error

chrisnas committed:
Deal with empty allocation context when fixing them

chrisnas committed:
Handle the case of objects larger than an allocation context

chrisnas committed:
Take review into account

chrisnas committed:
Take review into account for dynamically sampling

chrisnas committed:
Adding a first test that AllocationSampled is emitted when keyword/verbosity is set

chrisnas committed:
Check type name in AllocationSampled test

chrisnas committed:
Compare perf impact of allocation sampling

chrisnas committed:
Fix tests

chrisnas committed:
Update tests and doc

chrisnas committed:
Update review

chrisnas committed:
Update based on review

chrisnas committed:
Add tests and simple framework to measure statistical distribution of allocations

chrisnas committed:
Add x1/x2/x3 ratio check

chrisnas committed:
Update array of double implementation for 32 bits

chrisnas committed:
Updates

chrisnas committed:
Fix markdown

chrisnas committed:
Add AllocationsRunEventSource to compute percentiles more easily

chrisnas committed:
Allow percentiles computation

chrisnas committed:
Add ratio based array allocations

chrisnas committed:
take review into account

chrisnas committed:
Refactor the code and update upscaling method

chrisnas committed:
Add runs for allocations

chrisnas committed:
Handle padding for objects alignment
Use different upscaling

chrisnas committed:
take GC padding into account

chrisnas committed:
Update allocation sampling distribution results

chrisnas committed:
Update benchmarks with events

chrisnas committed:
Beginning of NativeAOT support

chrisnas committed:
Fix trailing spaces

chrisnas committed:
Updates based on PR comments

noahfalk committed:
Proposed changes

- Renamed alloc_sampling -> combined_limit
   We were refering to alloc_sampling as the sampling limit but really it is min(sampling_limit, alloc_limit).
   Most of the time alloc_sampling was the AC limit and not a sampling limit. It seemed misleading to
   have 'sampling' in the name when most of the time that isn't what it repreesnts.

- Simplified logic in Alloc()
   Special casing UOH allocations appears unnecessary, performance still looked good without doing that
   Factored all the checks to see if the feature was on inside a IsRandomizedSamplingEnabled() func
   Moved all the fields that are only needed for sampling inside the scope where sampling is enabled
   Adjusted some comments

-Fixed the sampled byte reporting for objects that extend past the AC boundary

- Removed unneeded code changes that were fragments of past code changes no longer being used

- Removed errant whitespace changes in gc code

- consolidated test changes in a folder named "randomizedallocationsampling". We don't need multiple
  folders and I'm trying to consolidate naming this feature as 'randomized' allocation rather than
  'dynamic'

- moved the testing results into a dedicated folder that can be deleted at the end to make it clear what
  stays and what is temporary just for review.

noahfalk committed:
Fix README

I left behind a random fragment, now deleted

chrisnas committed:
- fix Alloc helper to get the address of the allocated object for AllocationSampled
- revert initial changes to basic evenpipe test now that a dedicated one exists
- update comment in sample application to replace 100000 by 102400

noahfalk committed:
Update the design doc

-Fixed the example distribution of errors and other places that 1/100,000 was being used instead of 1/102400
-Changed alloc_sampling -> combined_limit
-Improved some of the statistics discussion

chrisnas committed:
Merge branch 'main' into chrisnas/alloc_sampling

noahfalk commented:
fyi @brianrob @Maoni0 @jkotas @davidwr @AaronRobinsonMSFT @dotnet/dotnet-diag 
At this point I'd expect minor changes addressing some CI failures for CoreCLR, and potentially larger changes as the NativeAOT support is completed. If you have any early feedback do let us know, and then we'll leave draft mode once CI is clean and NativeAOT is done.

noahfalk commented on src/coreclr/gc/gc.cpp:
```diff
@@ -44076,7 +44076,7 @@ size_t gc_heap::decommit_region (heap_segment* region, int bucket, int h_number)
 #ifdef MULTIPLE_HEAPS
         // In return_free_region, we set heap_segment_heap (region) to nullptr so we cannot use it here.
-        // but since all heaps share the same mark array we simply pick the 0th heap to use. 
+        // but since all heaps share the same mark array we simply pick the 0th heap to use.
```
Nit:
```suggestion
        // but since all heaps share the same mark array we simply pick the 0th heap to use. 
```

chrisnas replied:
✅ done

davidwr commented:
@davidwrighton

chrisnas committed:
Fix compilation errors for Linux and related to merged changes in Thread class

chrisnas committed:
Fix compilation issue

chrisnas committed:
Restore Thread randomizer initialization

noahfalk commented on src/coreclr/vm/threads.h:
```diff
@@ -3776,6 +3777,11 @@ class Thread
     // See ThreadStore::TriggerGCForDeadThreadsIfNecessary()
     bool m_fHasDeadThreadBeenConsideredForGCTrigger;
 
+    CLRRandom m_random;
```
This is reverting a change that was made here: https://github.com/dotnet/runtime/pull/99989/files#r1531477052

I suspect it will be worthwhile to have a thread-local random number generator to avoid lock contention in the Alloc() code path, but perhaps this should now be a pointer that is allocated on demand rather than eagerly allocated in the Thread object.
@jkotas - any thoughts since it looks like you were advocating for the original removal of this field?

noahfalk replied:
@jkoritzinsky - your thoughts as well?

jkotas replied:
I think it should be a free-standing static. There is no reason for all random statics to be centralized on the Thread type.

> this should now be a pointer that is allocated on demand rather than eagerly allocated in the Thread object.

Yes. Or switch to a more compact random implementation. 220 bytes of state for pseudo random number generator sounds like an overkill. E.g. [Xorshiro](https://github.com/dotnet/runtime/blob/ce1477b5df9b378f1dfcbb9194aeed3932ecfc03/src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs#L31) used by BCL currently has 16 bytes of state.

chrisnas committed:
Remove the randomizer from Thread to have just a singleton

jkotas commented on src/coreclr/vm/gcheaputilities.cpp:
```diff
@@ -41,7 +41,11 @@ bool g_sw_ww_enabled_for_gc_heap = false;
 
-GVAL_IMPL_INIT(gc_alloc_context, g_global_alloc_context, {});
+CLRRandom m_random; // used to randomize allocation sampling
+CLRRandom* GetRandomizer() { return &m_random;}
```
This does not look thread safe.

CLRRandom is not thread safe. Using it from multiple threads may corrupt its state and it can start returning degraded random numbers or even a constant.

chrisnas replied:
Good point...
However, I'm not sure that this PR is worth implementing [Xorshiro](https://github.com/dotnet/runtime/blob/ce1477b5df9b378f1dfcbb9194aeed3932ecfc03/src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs#L31) in C and have it per thread. Would it be possible to keep the original ClrRandom per thread?

jkotas replied:
ClrRandom per thread is fine. I should be allocated lazily (https://github.com/dotnet/runtime/pull/100356#discussion_r1614246680).

chrisnas committed:
Add a lazily allocated CLRRandom in Thread class
Still a few issues to solve:
  - where to delete the m_pRandom?
  - how to use Thread's members in gcheaputilities.h?

chrisnas commented on src/coreclr/vm/threads.h:
```diff
@@ -3777,6 +3777,13 @@ class Thread
+    CLRRandom* m_pRandom;
+
+    public:
+        // TODO: where to delete the allocated CLRRandom object?
```
I tried to find an example in the CLR code. The EventPipeThreadHolder* eventpipe_tls_instance is deleted when the free_thread_holder function is explicitly called. Same for t_pCurrentThreadLog that uses a different list mechanism but again, with a cleanup mechanism explicitly triggered.
I don't know when it is 100% sure that no code will allocate for a given thread. The allocation context is cleared in Thread::OnThreadTerminate. Do you confirm this could be the right place?

jkotas replied:
No code should call CLRRandom during thread shutdown, so the ordering should not matter.

It can be something like this:
```cpp
//  .h file;

class CLRRandom
{
    ....

private:
    struct CLRRandomHolder
    {
        CLRRandom* _p;

        CLRRandomHolder()
        {
            _p = new CLRRandom();
            _p->Init();
        }

        ~CLRRandomHolder()
        {
            delete _p;
        }
    };

    static thread_local CLRRandomHolder t_instance;

public:
    CLRRandom* GetSharedRandom()
    {
        return t_instance._p;
    }
};

---
// .cpp file

thread_local CLRRandom::CLRRandomHolder CLRRandom::t_instance = CLRRandomHolder();
---

chrisnas replied:
✅ done

chrisnas committed:
Start a no op implementation of randomized sampling for NativeAOT
A lot of code is commented due to compilation errors hard to
fix: what include file to use without triggering more compilation
errors.

brianrob commented on docs/design/features/RandomizedAllocationSampling.md:
```diff
@@ -0,0 +1,332 @@
+
+When the new AllocationSampled event is enabled, each managed thread starts sampling independent of one another. For a given thread there will be a sequence of allocated objects Object_1, Object_2, etc that may continue indefinitely. Each object has a corresponding .NET type and size. The size of an object includes the object header, method table, object fields, and trailing padding that aligns t ...
+
+When each new object is allocated, conceptually the runtime starts doing independent [Bernoulli Trials](https://en.wikipedia.org/wiki/Bernoulli_trial) (weighted coin flips) for every byte in the object. Each trial has probability p = 1/102,400 of being considered a success. As soon as one successful trial is observed no more trials are performed for that object and an AllocationSampled event is e ...
```
Typo: `trails` --> `trials`.

chrisnas replied:
✅ done

brianrob commented on src/coreclr/nativeaot/Runtime/GCHelpers.cpp:
```diff
@@ -471,6 +481,43 @@ EXTERN_C int64_t QCALLTYPE RhGetTotalAllocatedBytesPrecise()
+        gc_heap* hp = gc_heap::heap_of((BYTE*)orObject);
+        heapIndex = hp->heap_number;
+#endif
+        FireEtwAllocationSampled(allocKind, GetClrInstanceId(), typeId, name, heapIndex, (BYTE*)orObject, size, samplingBudgetOffset);
```
You may already have this on a list somewhere, but `name` is always `null` (I'm not sure that nativeaot even has the type name data available...)

chrisnas replied:
> You may already have this on a list somewhere, but `name` is always `null` (I'm not sure that nativeaot even has the type name data available...)

I'm reusing the same code as AllocationTick to get the type name and it is working in the tests for Core. I'll try the same code for NativeAOT when the compilations errors are fixed.

brianrob replied:
Sounds good.  I suspect that nativeaot doesn't even have the type names, so it may just always need to be null, but wanted to call it out.

chrisnas replied:
Sorry @brianrob!
I missed that your point was for NativeAOT only!

brianrob commented:
@chrisnas, @noahfalk I was thinking a bit about how to introduce this into collection tools.  Unfortunately, there isn't really a great way to query and see if this feature is supported ahead of time, and of course with machine-wide capture, we're highly likely to get a mix of answers.  I think it would make sense to have a way to turn on `AllocationSampled` events for processes that support it, but still capture `AllocationTick` events for those that don't.  Right now, that would mean enabling the `GC` keyword and the `AllocationSampling` keyword at the `verbose` level.  Unfortunately, this will mean that we capture both events for processes that support both.  What do you guys think about adding some code to check for this, and only emit `AllocationTick` events if `AllocationSampling` is not enabled?  Specifically:

 - If `AllocationSampled` is not supported, then it has no impact.
 - If `AllocationSampled` is supported but not enabled, no change to behavior of `AllocationTick`.
 - If `AllocationSampled` is supported and enabled, then `AllocationTick` is not emitted, even if requested.

If we really wanted to get them both, we could add an override keyword, but I'm inclined not to do this.  What do you think?

noahfalk commented:
> If we really wanted to get them both, we could add an override keyword, but I'm inclined not to do this. What do you think?

My initial gut reaction was that it feels weird to have keywords automatically turn off other events but the more I think on it that does seem like behavior profiler are likely to prefer. I'd need to look more closely to see what is involved.

chrisnas committed:
Fix typos in markdown

chrisnas committed:
Fix linux compilation errors

brianrob commented:
> > If we really wanted to get them both, we could add an override keyword, but I'm inclined not to do this. What do you think?
> 
> My initial gut reaction was that it feels weird to have keywords automatically turn off other events but the more I think on it that does seem like behavior profiler are likely to prefer. I'd need to look more closely to see what is involved.

Sounds good.  I think I went through the same mental thought process.

chrisnas committed:
Make alloc context randomizer per thread lazily allocated
+ revert SPACE change in gc.cpp

chrisnas committed:
Fix missing rename in assembly code to use combined_limit instead of alloc_limit

chrisnas commented:
> it would make sense to have a way to turn on `AllocationSampled` events for processes that support it, but still capture `AllocationTick` events for those that don't. Right now, that would mean enabling the `GC` keyword and the `AllocationSampling` keyword at the `verbose` level. Unfortunately, this will mean that we capture both events for processes that support both.

Since we don't have a way to "activate" only one event but instead all events under a keyword/verbosity, how could we know that gc/verbose means AllocationTick only and not other GC events? I mean that it should be possible (but I don't where in the CLR code) to "disable" Allocation tick if AllocationSampling/informal is activated but the other GC events will still be emitted. I'm not sure of the gain in term of performance. I'm expecting Perfview to have a new checkbox for .NET 9+ randomized allocation sampling that the user would check.
Note that, in a profiler, we know which version of the runtime is loading us so it should not be a problem to pick one or the other based on runtime version > 9.0.

brianrob commented:
> > it would make sense to have a way to turn on `AllocationSampled` events for processes that support it, but still capture `AllocationTick` events for those that don't. Right now, that would mean enabling the `GC` keyword and the `AllocationSampling` keyword at the `verbose` level. Unfortunately, this will mean that we capture both events for processes that support both.
> 
> Since we don't have a way to "activate" only one event but instead all events under a keyword/verbosity, how could we know that gc/verbose means AllocationTick only and not other GC events? I mean that it should be possible (but I don't where in the CLR code) to "disable" Allocation tick if AllocationSampling/informal is activated but the other GC events will still be emitted. I'm not sure of the gain in term of performance. I'm expecting Perfview to have a new checkbox for .NET 9+ randomized allocation sampling that the user would check. Note that, in a profiler, we know which version of the runtime is loading us so it should not be a problem to pick one or the other based on runtime version > 9.0.

I think the way that you could detect this is in [fire_etw_allocation_event](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gcee.cpp#L306).  You can wrap the call to `FIRE_EVENT` in  `if(EVENT_ENABLED(GCAllocationTick) && !EVENT_ENABLED(GCAllocationSampled))`.

jkotas commented on src/coreclr/vm/gchelpers.cpp:
```diff
@@ -424,54 +530,12 @@ OBJECTREF AllocateSzArray(MethodTable* pArrayMT, INT32 cElements, GC_ALLOC_FLAGS
         else
-#endif  // FEATURE_64BIT_ALIGNMENT
         {
 #ifdef FEATURE_64BIT_ALIGNMENT
```
```suggestion
#ifdef FEATURE_DOUBLE_ALIGNMENT_HINT
            if (pArrayMT->GetArrayElementType() == ELEMENT_TYPE_R8)
            {
                flags |= GC_ALLOC_ALIGN8;
            }
#endif
#ifdef FEATURE_64BIT_ALIGNMENT
```
And delete the `if ((DATA_ALIGNMENT < sizeof(double)) && (pArrayMT->GetArrayElementType() == ELEMENT_TYPE_R8))` check above.

chrisnas replied:
So it means that there is only one code path without any check for FEATURE_64BIT_ALIGNMENT now?
``` 
else
{
#ifdef FEATURE_DOUBLE_ALIGNMENT_HINT
   if (pArrayMT->GetArrayElementType() == ELEMENT_TYPE_R8)
   {
      flags |= GC_ALLOC_ALIGN8;
   }
#endif
   orArray = (ArrayBase*)Alloc(totalSize, flags);
   orArray->SetMethodTable(pArrayMT);
}
```

jkotas replied:
I meant to make it look like this:
```
#ifdef FEATURE_DOUBLE_ALIGNMENT_HINT
   if (pArrayMT->GetArrayElementType() == ELEMENT_TYPE_R8)
   {
      flags |= GC_ALLOC_ALIGN8;
   }
#endif
#ifdef FEATURE_64BIT_ALIGNMENT
    MethodTable* pElementMT = pArrayMT->GetArrayElementTypeHandle().GetMethodTable();
    if (pElementMT->RequiresAlign8() && pElementMT->IsValueType())
    {
        // This platform requires that certain fields are 8-byte aligned (and the runtime doesn't provide
        // this guarantee implicitly, e.g. on 32-bit platforms). Since it's the array payload, not the
        // header that requires alignment we need to be careful. However it just so happens that all the
        // cases we care about (single and multi-dim arrays of value types) have an even number of DWORDs
        // in their headers so the alignment requirements for the header and the payload are the same.
        _ASSERTE(((pArrayMT->GetBaseSize() - SIZEOF_OBJHEADER) & 7) == 0);
        flags |= GC_ALLOC_ALIGN8;
    }
#endif
   orArray = (ArrayBase*)Alloc(totalSize, flags);
   orArray->SetMethodTable(pArrayMT);
```

chrisnas replied:
Make more sense  :^)
✅ done

chrisnas commented:
> I think the way that you could detect this is in [fire_etw_allocation_event](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gcee.cpp#L306). You can wrap the call to `FIRE_EVENT` in `if(EVENT_ENABLED(GCAllocationTick) && !EVENT_ENABLED(GCAllocationSampled))`.

I was thinking checking a step earlier in [gc_heap::update_alloc_info](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gc.cpp#L18846) so that the etw_allocation_running_amount[oh_index] would not be updated and exceeded_p would be false if EVENT_ENABLED(GCAllocationSampled).

I looked at ClrEtwAll.man and, in addition to AllocationTick, a few GC events are triggered when verbose: FinalizeObject, PinObjectAtGCTime, IncreaseMemoryPressure, DecreaseMemoryPressure, GCJoin (this one could be noisy if I remember well some of my tests), and GCFitBucketInfo. So these events would be emitted even if AllocationTick is filtered out

chrisnas committed:
Fix build errors dues to explicit net8.0 usage in tests

brianrob commented:
> > I think the way that you could detect this is in [fire_etw_allocation_event](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gcee.cpp#L306). You can wrap the call to `FIRE_EVENT` in `if(EVENT_ENABLED(GCAllocationTick) && !EVENT_ENABLED(GCAllocationSampled))`.
> 
> I was thinking checking a step earlier in [gc_heap::update_alloc_info](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gc.cpp#L18846) so that the `etw_allocation_running_amount[oh_index]` would not be updated and `exceeded_p` would be false if `EVENT_ENABLED(GCAllocationSampled)`.
> 
> I looked at ClrEtwAll.man and, in addition to **AllocationTick**, a few GC events are triggered when verbose: **FinalizeObject**, **PinObjectAtGCTime**, **IncreaseMemoryPressure**, **DecreaseMemoryPressure**, **GCJoin** (this one could be noisy if I remember well some of my tests), and **GCFitBucketInfo**. So these events would be emitted even if **AllocationTick** is filtered out
> 
> BTW, the EVENT_ENABLED macro is based on internal GC helpers and since AllocationSampled is not a GC event, it does not appear in [gcevents.h](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gcevents.h#L24) like AllocationTick. Also these macros are based on 2 different GC providers + a [dedicated implementation for enabled/disabled state](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gceventstatus.h#L63). However, it should be possible to use
> 
> ```
> ETW_TRACING_CATEGORY_ENABLED(
>    MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context,
>    TRACE_LEVEL_INFORMATION,
>    CLR_ALLOCATIONSAMPLING_KEYWORD)
> ```
> 
> that is now usabled from Linux thanks to @noahfalk ;^)
> 
> **NOTE:** After testing, it seems that these macros are defined in eventtracebase.h and when I'm trying to #include it in gc.cpp, a lot of compilation errors are triggered.

I think it would be reasonable to address this either way.  With regard to `ETW_TRACING_CATEGORY_ENABLED`, you're right that you can't call this from within the GC.  I *think* you should be able to declare the event as a `KNOWN_EVENT` at which point the GC will know about it.  You don't have to emit the event from the GC, but the GC will then know whether or not the event has been enabled.

Maoni0 commented:
`update_alloc_info` doesn't only update the event related stuff so I would do it the way Brian suggested. 

please do make sure this works with the standalone GC. it's clrgcexp.dll. 

I haven't been feeling well so I haven't had a chance to take a look (but @noahfalk had already talked to me about the changes in GC) - I'll try to look at this soon.

chrisnas commented:
> `update_alloc_info` doesn't only update the event related stuff so I would do it the way Brian suggested.
I was thinking about a scenario where allocationtick could be enabled AFTER a trace where AllocationSampled would have been enabled and then ended. In that case, I thought that specific GC increment counters should not be udated like in the following code:
```
inline
bool gc_heap::update_alloc_info (int gen_number, size_t allocated_size, size_t* etw_allocation_amount)
{
    bool exceeded_p = false;
    int oh_index = gen_to_oh (gen_number);
    allocated_since_last_gc[oh_index] += allocated_size;

    if (EVENT_ENABLED(AllocationSampled))
    {
        // don't emit AllocationTick if AllocationSampled is enabled
        exceeded_p = false;
    }
    else
    {
        size_t& etw_allocated = etw_allocation_running_amount[oh_index];
        etw_allocated += allocated_size;
        if (etw_allocated > etw_allocation_tick)
        {
            *etw_allocation_amount = etw_allocated;
            exceeded_p = true;
            etw_allocated = 0;
        }
    }

    return exceeded_p;
}
```
But no problem to disable AllocationTick where Brian mentioned

> please do make sure this works with the standalone GC. it's clrgcexp.dll.
I will
 
> I haven't been feeling well so I haven't had a chance to take a look (but @noahfalk had already talked to me about the changes in GC) - I'll try to look at this soon.
Take the time to get better  :^)

chrisnas commented:
> I _think_ you should be able to declare the event as a `KNOWN_EVENT` at which point the GC will know about it. You don't have to emit the event from the GC, but the GC will then know whether or not the event has been enabled.

Yes: I'm updating gcinterface.h to define the keyword, make the enum uint64_t to support its large value and define the corresponding event in gcevents.h with KNOWN_EVENT.

One last thought about that filtering: it won't be possible any more to compare AllocationTick and AllocationSampled in the same execution run like I'm doing in the test application.

chrisnas committed:
Filter out AllocationTick when AllocationSampled is enabled

chrisnas committed:
Try a fix for build errors related to .csproj that are not really tests

chrisnas committed:
Avoid touching the IGCToCLREventSink interface because AllocationSampled
do not need to be emitted

jkotas commented on src/coreclr/nativeaot/Runtime/thread.h:
```diff
@@ -99,6 +101,31 @@ struct ThreadBuffer
+    //   This occurs either because the sampling feature is disabled, or because the randomized selection
+    //   of sampled bytes didn't select a byte in this AC.
+    // - if combined_limit < alloc_limit there is a sample limit in the AC. sample_limit = combined_limit.
+    uint8_t* m_combined_limit;
```
This should be in front of `m_rgbAllocContextBuffer` to make it as similar to regular CoreCLR as possible, and to improve locality of data used by the allocation helpers.

chrisnas replied:
Good point.
I was thinking about it at the beginning but I did not want to update all the offsets for the assembly code before ensuring that the rest was working. I will definitively do it when everything compiles and run like a charm  :^)

chrisnas replied:
✅ done

jkotas commented on src/coreclr/gc/gcinterface.h:
```diff
@@ -265,7 +265,7 @@ enum GCEventLevel
 // numbers come from the ETW manifest itself - please make changes to this enum
 // if you add, remove, or change keyword sets that are used by the GC!
-enum GCEventKeyword
+enum GCEventKeyword : uint64_t
```
This enum is used by the GC/EE interface. Changing the size of the enum to `uint64_t` would required revving `GC_INTERFACE_MAJOR_VERSION`

jkotas replied:
I expect that @Maoni0 would prefer to avoid revving the GC interface major version and do this in some other way.

chrisnas replied:
@noahfalk @brianrob: the current implementation of the provider/keyword/verbosity status update in [EtwCallbackCommon](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/eventtrace.cpp#L2400) casts the 64 bit received keyword into a 32 bit enum for the GC. Unfortunately, the AllocationSamplingKeyword is beyond this range.
I did not realize (thanks for pointing this out @jkotas) that it would require to change the signature of the GC interface to support it.
Any idea?

brianrob replied:
So far, I'm not seeing something that I would recommend.  I can think of at least one way to do this, but I don't even want to say it because I think it could break us into eventual jail.  The idea would be to use a bit that is already used on the ETW keyword mask, but not sent to the GC because the GC doesn't currently use it.  The risk is that if the GC ever wanted to use said bit, then there would be a problem.

So far, revving the interface seems best to me.

noahfalk replied:
I think we could just do the filtering on the VM side in the GCToCLREventSink no?
https://github.com/dotnet/runtime/blob/f6701e58766c8556919bc92cf2f9b33b6bb2383b/src/coreclr/vm/gctoclreventsink.cpp#L163

brianrob replied:
Yes, that would be fine.  I had been trying to avoid the interface call cost, but I think it will not be a big deal, and certainly more preferrable to changing the interface.

chrisnas replied:
By following the execution path of `GCHeapUtilities::RecordEventStateChange`, it should be possible to call a new `GCHeapUtilities::AllocationSampledEventState(bool enable)` based on the value of the received keyword. This would be non generic processing like what is done for triggering a full GC. 
On the GC side, a simple boolean field in `GCEventStatus` would keep track of the state and be exposed like in my current code via `GCEventEnabledAllocationSampled()`.
This would also allow enable/disable detection that does not seem possible today (once enabled, a provider/keyword/verbosity stays enabled) and... avoid the virtual call  ;^)

brianrob commented:
> > I _think_ you should be able to declare the event as a `KNOWN_EVENT` at which point the GC will know about it. You don't have to emit the event from the GC, but the GC will then know whether or not the event has been enabled.
> 
> Yes: I'm updating gcinterface.h to define the keyword, make the enum uint64_t to support its large value and define the corresponding event in gcevents.h with KNOWN_EVENT.
> 
> One last thought about that filtering: it won't be possible any more to compare AllocationTick and AllocationSampled in the same execution run like I'm doing in the test application.

You're right.  If we think that this is really necessary, we could add an override keyword, but I'm inclined to not add one unless we have a scenario.  What do you think?

noahfalk commented:
> > One last thought about that filtering: it won't be possible any more to compare AllocationTick and AllocationSampled in the same execution run like I'm doing in the test application.

> You're right. If we think that this is really necessary, we could add an override keyword, but I'm inclined to not add one unless we have a scenario. What do you think?

An override keyword does not seem necessary to me. We can measure AllocationTick in one run to AllocationSampled in a 2nd run. This is the same comparison any profiler is going to observe if they switch between GC-verbose-only and GC-verbose + AllocationSampled.

chrisnas committed:
Update AllocationTick filtering implementation

chrisnas committed:
Fix double[] implementation

chrisnas committed:
Fix test due to added field in the event payload

chrisnas committed:
Implement emitting the AllocationSampled event in NativeAOT

chrisnas committed:
First no op implementation for NativeAOT with updated assembly code

noahfalk commented on src/coreclr/gc/gcinterface.h:
```diff
@@ -280,6 +280,7 @@ enum GCEventKeyword
     GCEventKeyword_ManagedHeapCollect            =  0x800000,
     GCEventKeyword_GCHeapAndTypeNames            = 0x1000000,
     GCEventKeyword_GCSampledObjectAllocationLow  = 0x2000000,
+
```
Nit: revert whitespace changes

chrisnas replied:
✅ done

noahfalk commented on src/coreclr/nativeaot/Runtime/eventtracebase.h:
```diff
@@ -103,6 +103,9 @@ struct ProfilingScanContext;
 #define CLR_MANAGEDHEAPCOLLECT_KEYWORD 0x800000
 #define CLR_GCHEAPANDTYPENAMES_KEYWORD 0x1000000
 
+// TODO: why not using the constants from ClrEtwAll.h?
```
I don't know, but I wouldn't worry about it and follow the existing pattern

chrisnas replied:
✅ done

chrisnas committed:
Fix typos

chrisnas committed:
Fix DEBUG compilation error

chrisnas committed:
Fix linux/x86 compilation errors

chrisnas committed:
Fix build error

chrisnas committed:
Try to fix "static" related compilation errors

noahfalk commented:
> [@chrisnas] Following [the doc](https://github.com/dotnet/runtime/blob/main/docs/design/features/standalone-gc-loading.md#identifying-candidate-shared-libraries), I tried to run my tests with this standalone GC on a .NET 9 branch without my changes:

> ...

> I set DOTNET_GCName=...\runtime\artifacts\tests\coreclr\windows.x64.Release\Tests\Core_Root\clrgcexp.dll


Unforetunately the info in that old document is out-of-date as the configuration has changed slightly. This public docs page has the up-to-date information: https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#standalone-gc

Instead of putting a path in the DOTNET_GCName env var you should just put the file name. For example:
DOTNET_GCName=clrgcexp.dll

I built with the current changes in this PR and ran with the env var set that way, did not encounter any crashes/initialization issues, and confirmed with a debugger that it really was loading the standalone GC dll. I'd recommend doing it yourself just to double check and validate that all the event generation still works as expected, but pretty sure its going to be fine.

Maoni0 commented:
do you have perf results that compare alloctick vs the sampling events with everything being the same, ie, with GC events on at verbose level?

Maoni0 commented:
also I did look at the changes in GC and they look good to me

chrisnas commented:
> do you have perf results that compare alloctick vs the sampling events with everything being the same, ie, with GC events on at verbose level?

I will compare now that AllocationTick is filtered out
You can find the current results (AllocationTick vs AllocationSampled) in src/tests/tracing/eventpipe/randomizedallocationsampling/manual/testing_results/perf/GCPerfSimx10_Baseline+AllocationTick.txt and src/tests/tracing/eventpipe/randomizedallocationsampling/manual/testing_results/perf/GCPerfSimx10_PullRequest+Events.txt

chrisnas committed:
Fix test in 32 bit

chrisnas commented:
> (@noahfalk) I'd recommend doing it yourself just to double check and validate that all the event generation still works as expected, but pretty sure its going to be fine.

I've checked that when clrgcexp.dll is loaded:
- AllocationTick is still emitted if AllocationSampled is not enabled
- AllocationSampled is emitted when enabled
- AllocationTick is not emitted when AllocationSampled is enabled

chrisnas commented:
> do you have perf results that compare alloctick vs the sampling events with everything being the same, ie, with GC events on at verbose level?

Here are the results @Maoni0: 
```
	ALL EVENTS	ALLOCATIONTICK	ALLOCATIONSAMPLED
--------------------------------------------------------------------
average	22.83501899	21.96691761	21.4408235
median	22.84194655	21.8062514	21.14077555
```

noahfalk commented:
> Here are the results @Maoni0:

I'm not sure how to interpret your columns @chrisnas? I was expecting two columns only:
1. GC verbose events including AllocationTick (GC keyword + verbose level)
2. GC verbose events - AllocationTick + AllocSampling (GC keyword + AllocSampling keyword + verbose level)

chrisnas commented:
> I'm not sure how to interpret your columns @chrisnas? I was expecting two columns only:
> 
> 1. GC verbose events including AllocationTick events (GC keyword + verbose level)
> 2. GC verbose events - AllocationTick events + AllocSampling events (GC keyword + AllocSampling keyword + verbose level)

**ALL EVENTS** = (GC keyword + verbose level + AllocSampling keyword) --> should be your 2.
**ALLOCATIONTICK** = (GC keyword + verbose level) --> should be your 1.
**ALLOCATIONSAMPLED** = (informational level + AllocSampling keyword) --> just to have a "reference" of the new event only.

It could be interesting to get GC keyword + informational level  (= GC only events) + AllocSampling keyword because this is what I'm expecting people will mostly use during monitoring or investigations

chrisnas committed:
Avoid running AllocationSampled test with Mono + Fix compilation issue

chrisnas committed:
Start to implement the randomizer with existing 32 bit random

chrisnas committed:
Fix compilation error

chrisnas committed:
Allow eventing functions to be called from thread.cpp

chrisnas committed:
Merge branch 'main' into chrisnas/alloc_sampling

chrisnas committed:
Fix assert in AOT

chrisnas committed:
Fix link issue

jkotas commented on src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt:
```diff
@@ -28,9 +28,11 @@ endif (CLR_CMAKE_TARGET_WIN32)
 
 add_library(Runtime.WorkstationGC STATIC ${COMMON_RUNTIME_SOURCES} ${FULL_RUNTIME_SOURCES} ${RUNTIME_SOURCES_ARCH_ASM} ${RUNTIME_ARCH_ASM_OBJECTS})
 add_dependencies(Runtime.WorkstationGC aot_eventing_headers)
+add_dependencies(Runtime.WorkstationGC eventpipe-enabled)
```
Why do we need this dependency?

chrisnas replied:
> Why do we need this dependency?

To fix the error listed in https://github.com/dotnet/runtime/pull/100356/checks?check_run_id=26327770979

jkotas replied:
This should be fixed by adding a dummy implementation to src\coreclr\nativeaot\Runtime\disabledeventtrace.cpp.

The tracing support is optional in native AOT. We build two static libraries `eventpipe-enabled` and `eventpipe-disabled`. Only one of these libraries should be linked in into the final binary, depending on whether the tracing support is on or off.

chrisnas replied:
Thanks @jkotas: it makes much more sense than what I did...
I updated the PR accordingly

chrisnas committed:
Fix NativeAOT build

chrisnas committed:
Merge branch 'main' into chrisnas/alloc_sampling

chrisnas committed:
Fix missing references to gc_alloc_context

chrisnas committed:
Fix compilation issues

chrisnas commented:
@jkoritzinsky: could you have a look at this PR?
It seems that https://github.com/dotnet/runtime/pull/103055 and https://github.com/dotnet/runtime/pull/103607 have removed the allocation context from the Thread class but also removed the assembly code that was using alloc_limit to figure out if an allocation slow path should be triggered.
In this PR, we rely on a new field to trigger a "fake" slow path to be able to generate a better random distribution of allocation notification.
Could you provide more details about how we could use your new implementation (especially where are the fast assembly helpers) to get the same "hooking" mechanism?
Thanks in advance  :^)

jkotas commented:
> Could you provide more details about how we could use your new implementation (especially where are the fast assembly helpers) to get the same "hooking" mechanism?

The assembly helpers were replaced by the portable C/C++ helpers in jithelpers.cpp that you are handling in this PR already. There was no point in having these assembly helpers anymore - the code generated by the C/C++ compiler was identical.

chrisnas committed:
Update x86 implementation

chrisnas commented:
> > Could you provide more details about how we could use your new implementation (especially where are the fast assembly helpers) to get the same "hooking" mechanism?
> 
> The assembly helpers were replaced by the portable C/C++ helpers in jithelpers.cpp that you are handling in this PR already. There was no point in having these assembly helpers anymore - the code generated by the C/C++ compiler was identical.

This seems to be the case for x64 but not for x86.
There is still assembly code generated in jitinterfacex86.cpp that relies on stublinker.cpp (I'm just guessing here). I've pushed a commit to "fix" it but I would be very interested in getting your feedbacks @jkotas; especially for the code in X86EmitCurrentThreadAllocContextFetch() - should GetAllocContextHelper() return the gc_alloc_context or ee_alloc_context.
Excuse my lack of confidence in my assembly skills and how code gets generated in the CLR  :^)

jkotas commented on src/coreclr/vm/i386/stublinkerx86.cpp:
```diff
@@ -2434,7 +2434,10 @@ namespace
     gc_alloc_context* STDCALL GetAllocContextHelper()
     {
-        return &t_runtime_thread_locals.alloc_context;
+        // TODO: I'm not sure to understand if it should return the gc_alloc_context or the ee_alloc_context
```
It needs to return `ee_alloc_context` to be in sync with the other path in X86EmitCurrentThreadAllocContextFetch

chrisnas replied:
> It needs to return `ee_alloc_context` to be in sync with the other path in X86EmitCurrentThreadAllocContextFetch

Thanks: I updated the code
I still don't understand the end of the function:
```
    _ASSERTE(Thread::GetOffsetOfThreadStatic(&t_runtime_thread_locals.alloc_context) < INT_MAX);
    X86EmitAddReg(dstreg, (int32_t)Thread::GetOffsetOfThreadStatic(&t_runtime_thread_locals.alloc_context));
```
Why is it needed to call GetAllocContextHelper before since it should return the same thing. 
Am I missing something obvious?  :^)

jkotas replied:
The method has two orthogonal paths:

https://github.com/dotnet/runtime/blob/0746dd3d07cbecea031dab391a0dbbb657ba20e7/src/coreclr/vm/i386/stublinkerx86.cpp#L2456-L2464 is used on Unix. It calls into the helper to fetch the thread static.

https://github.com/dotnet/runtime/blob/0746dd3d07cbecea031dab391a0dbbb657ba20e7/src/coreclr/vm/i386/stublinkerx86.cpp#L2488-L2494 is used on Windows . It has the thread static fetch inline.

jkotas commented:
> Also, the NativeAOT implementation seems to still have assembly helpers (that are updated in this PR): is it expected?

Yes.

chrisnas committed:
Fix

chrisnas committed:
Fixes

chrisnas committed:
Merge remote-tracking branch 'origin' into chrisnas/alloc_sampling

chrisnas committed:
Fix rebase

chrisnas committed:
Fix linux gcc build issues

chrisnas committed:
Add missing alloc context fixup and randomizer code

chrisnas committed:
Merge remote-tracking branch 'origin' into chrisnas/alloc_sampling

jkotas commented:
You may consider peeling off some of the changes into separate PR to make this easier to land. For example, the changes in the GC proper around the 64-bit aligned allocs can be peeled off into a separate PR.

chrisnas commented:
> You may consider peeling off some of the changes into separate PR to make this easier to land. For example, the changes in the GC proper around the 64-bit aligned allocs can be peeled off into a separate PR.

Split into how many pieces @jkotas?
- 64-aligned allocs
- Core implementation
- NativeAOT implementation
- ???

jkotas commented:
I think the following split may work better:
- 64-aligned allocs
- Introduce the secondary allocation limit (for both CoreCLR and NAOT)
- Implement the actual sampling event

Smaller PRs are easier to stabilize, review and merge that locks in forward progress.

chrisnas committed:
Fix issue when TLS for alloc context is not initialized

chrisnas committed:
Merge remote-tracking branch 'origin' into chrisnas/alloc_sampling

chrisnas commented:
> I think the following split may work better:
> 
> * 64-aligned allocs
> * Introduce the secondary allocation limit (for both CoreCLR and NAOT)
> * Implement the actual sampling event
> 
> Smaller PRs are easier to stabilize, review and merge that locks in forward progress.

I'm not sure to understand the difference between the last 2 items: emitting the event is a tiny part of the work; the second item is the big one. Also, without the sampling itself, the code will be different (mostly what I did at the beginning of NAOT when I did not want to change the behaviour - combined_limit = alloc_limit)

chrisnas committed:
Change randomizer to use sxoshiro128++ for better statistical distribution

chrisnas committed:
Fix possible crash in GCInterface_GetTotalAllocatedBytesPrecise

jkotas commented:
> I'm not sure to understand the difference between the last 2 items: emitting the event is a tiny part of the work; the second item is the big one. 

The second item (Introduce the secondary allocation limit (for both CoreCLR and NAOT)) would be a mechanical change. It does not need any tests. It should not introduce any observable behavior changes. It is just touching a lot of different places and it is easy to miss a few of those somewhere.

I do not have strong opinions about the best way to split this, but I think this will need to split into multiple PRs to land it - giving how long it takes to stabilize it.

jkotas commented on src/coreclr/nativeaot/Runtime/xoshiro128plusplus.h:
```diff
@@ -0,0 +1,131 @@
+#pragma once
```
Should this be move to src\native\minipal and CoreCLR impl switched to it as well?

jkotas commented:
In any case, if you agree that the 64-aligned allocs cleanup would be a good independent PR, we can start with that.

chrisnas committed:
Fix possible desync between gc_alloc_context and combined_limit

chrisnas committed:
Merge remote-tracking branch 'origin' into chrisnas/alloc_sampling

jkotas commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -1678,13 +1679,15 @@ HCIMPL1_RAW(Object*, JIT_NewS_MP_FastPortable, CORINFO_CLASS_HANDLE typeHnd_)
-    _ASSERTE(allocPtr <= allocContext->alloc_limit);
-    if (size > static_cast<SIZE_T>(allocContext->alloc_limit - allocPtr))
+    _ASSERTE(allocPtr <= eeAllocContext->combined_limit);
+    if ((allocPtr == nullptr) || (size > static_cast<SIZE_T>(eeAllocContext->combined_limit - allocPtr)))
```
We do not want to adding extra checks to the allocation helper fast paths. It would cause undesirable performance regressions.

tommcdon commented:
@noahfalk I believe this PR is superseded by: 
* https://github.com/dotnet/runtime/pull/104849
* https://github.com/dotnet/runtime/pull/104851
* https://github.com/dotnet/runtime/pull/104955

If yes, @chrisnas @noahfalk any objections if we close this PR?

noahfalk commented:
> @noahfalk I believe this PR is superseded by:

> https://github.com/dotnet/runtime/pull/104849
https://github.com/dotnet/runtime/pull/104851
https://github.com/dotnet/runtime/pull/104955

Yep, thats correct. Closing this PR.
